### ISSUE TITLE
"Too many components"

### ISSUE THREAD

**Issue by razblack** [JimBobSquarePants/ImageProcessor#444]
I'm still using a version I pulled from May/2016, as I have not yet updated to .NET Core V1 (still using RC1, Update 2).. anyway,  I plan to update, but cringing on it at the moment.

However, I have run across a JPG type that is throwing this ImageFormatException "Too many components".

I'm waiting to get a hold of a copy of the image for testing now.. but just wanted to ask how rare does this occur?

Could exporting the image with a different software package (like Photoshop) make a difference on the resulting JPG image?

**Comment by JimBobSquarePants** [comment]
It's probably CMYK. The current jpeg decoder doesn't support that but the original Go decoder we have ported does. Sadly I can't get my head around Go at all so I'm at a loss to convert it.

The expert on the manner is @mweber26 who kindly ported the code across. I have done some further work in my Core-Flava branch to handle JFIF headers better but CMYK support is still missing. Perhaps once Flava becomes the main branch we can collaborate add the missing functionality?

**Comment by razblack** [comment]
Thanks m8, I'm still waiting on getting a hold of the image >.<, I'll post back here once I get it and take a looksie.

**Comment by mweber26** [comment]
Post the image here and I'll take a look at it.

**Comment by JimBobSquarePants** [comment]
Thanks @mweber26

**Comment by razblack** [comment]
@mweber26 ... if I can ever get them to send it!  >.<

I'm 99% sure though that @JimBobSquarePants was spot on...  based on the sketchy server logs I maintained (only dumped e.Messages on the exceptions), I generated a few CMYK JPG images in Photoshop and sure enough, "Too many components" was being thrown.

I will still post the image(s) that they send me though...

**Comment by JimBobSquarePants** [comment]
That's everything merged into the Core branch now so once we have a test image we are good to go.

**Comment by JimBobSquarePants** [comment]
![imageprocessor-logo-512](https://cloud.githubusercontent.com/assets/385879/17885084/e220b114-695e-11e6-8bb2-f533f022144e.jpg)
Hi @mweber26 This image should be CMYK for debugging against. Could you have a wee look when you can please?

**Comment by mweber26** [comment]
I will try to look at fixing this, but it looks like you guys have moved on to .Net core and I don't use that (under linux and mac).  So I can't get the current master to compile.  I will try to fix under an older version and get a diff.

**Comment by mweber26** [comment]
Like I said -- I can't really make a pull request since I can't compile the master branch.  But here is the diff from an old revision to enable CMYK (4 components)

https://github.com/mweber26/ImageProcessor/commit/dfabf30fa872444cff7b0f427850b031190461e1

**Comment by JimBobSquarePants** [comment]
That's absolutely brilliant @mweber26 Thanks for your help! 

I was able to easily translate your work to a solution I could include it. There's been a lot of changes in the repo since you last looked. 

I dunno what the tooling is like on linux and mac for .NET Core as it stands. That's our target now as it means much greater visibility

**Comment by mweber26** [comment]
No problem.  I understand.  The tooling for .Net Core on linux and mac is very good (I've looked at it before).  We just haven't transitioned to it at our company since we have third party libraries that have not made the transition.

I tried to compile Core on standard .net mono and the thing that was incompatible was the System.Runtime.CompilerServices.Unsafe class (from PixelAccessor), which only exists in .net core.  There could be a way to #if for the two versions.  If we ever need the new features, I can look into that solution.

**Comment by JimBobSquarePants** [comment]
Yeah, there's a lot of stuff still to make the transition. We're pioneers here! 

I'm very surprised that wasn't able to compile. The support matrix seems quite large according to [nuget](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/), I actually thought it would be complete given that it is a single class made up of IL. 

Unfortunately no way around it just now. An alternative would have required multiple classes to cater for each `IPackedVector<T>` implementation which gets messy fast.

**Comment by GarageWeb** [comment]
Hello, Thanks for the fix. Are you going to update the NuGet packages to include the fixes since the initial alpha release?

**Comment by GarageWeb** [comment]
I may not understand how CMYK jpegs are processed here, so forgive me if this is not the best question. It seems like whenever I read a CMYK image stream into an IMAGE object, the height and width are both zero. 
example: imageBuffer is a stream from a CMYK image that I am reading from a URL. It is a valid stream and has the right data. 

When I do this: var original = new Image(imageBuffer);

original always has a height of zero and a width of zero. This works perfectly fine with normal rgb jpegs

**Comment by frarodnie** [comment]
Hello, I'm using last library version 2.5.3, I did a little test just opening a JPG image in CMYK with ImageFactory and then just saving to check if preserve colours, comparing both original and new it is noticiable difference in colours. Is there a way to preserve the original colours?.  Thanks.

**Comment by JimBobSquarePants** [comment]
I'm afraid not. I'm limited by the System.Drawing decoder. We're working on adding ICC profile support to ImageSharp though.
